Additional docs for shared_cache searchable snapshots (#70566)
elastic/elasticsearch: docs/reference/searchable-snapshots/index.asciidoc

[[searchable-snapshots]]
== {search-snaps-cap}

{search-snaps-cap} let you use <<snapshot-restore,snapshots>> to search and
store infrequently-accessed, read-only data. The <<cold-tier,cold>> and
<<frozen-tier,frozen>> data tiers use {search-snaps} to reduce your storage and
operating costs.

{search-snaps-cap} eliminate the need for <<scalability,replica shards>>,
potentially halving the local storage needed to search your data.
{search-snaps-cap} rely on the same snapshot mechanism you already use for
backups and have minimal impact on your snapshot repository storage costs.

[discrete]
[[mount-options]]
=== Mount options

To search a snapshot, you must first mount it locally as an index. There are two
mounting options, each with a different local storage footprint:

[[full-copy]]
Full copy::
Loads a full, local copy of the snapshotted index's shards into the cluster. The

> cold tier uses this option by default.

Reviewer: Does the hot tier also use this option by default? I was wondering what was used for searchable snapshots created during the hot phase.

[[searchable-snapshots]]
== {search-snaps-cap}

{search-snaps-cap} let you use <<snapshot-restore,snapshots>> to search and
store infrequently accessed, read-only data. The <<cold-tier,cold>> and
<<frozen-tier,frozen>> data tiers use {search-snaps} to reduce your storage and
operating costs.

{search-snaps-cap} eliminate the need for <<scalability,replica shards>>,
potentially halving the local storage needed to search your data.
{search-snaps-cap} rely on the same snapshot mechanism you already use for
backups and have minimal impact on your snapshot repository storage costs.

[discrete]
[[mount-options]]

> === Mount options

Reviewer 2: I would rather move this section below the "using searchable snapshots" section, maybe integrating it with "how searchable snapshots work" better, because we prefer users to use ILM for getting data into searchable snapshots and I think having this detail come first will guide users in the wrong direction.

<<searchable-snapshot-mount-storage-options,mount options>> specified. If
possible, searches use data from local storage. If the data is not available
locally, {es} downloads the data that it needs from the snapshot repository.

If a node holding one of these shards fails, {es} automatically allocates the
affected shards on another node, and that node restores the relevant shard data
from the repository. No replicas are needed, and no complicated monitoring or
orchestration is necessary to restore lost shards. Although searchable snapshot
indices have no replicas by default, you may add replicas to these indices by
adjusting `index.number_of_replicas`. Replicas of {search-snap} shards are
recovered by copying data from the snapshot repository, just like primaries of
{search-snap} shards. In contrast, replicas of regular indices are restored by
copying data from the primary.

[discrete]
[[searchable-snapshot-mount-storage-options]]
==== Mount options

To search a snapshot, you must first mount it locally as an index. Usually
{ilm-init} will do this automatically, but you can also call the
<<searchable-snapshots-api-mount-snapshot,mount snapshot>> API yourself. There

> are two possibilities for the `?storage` mount option, each with different
> performance characteristics and a different local storage footprint:

Reviewer: Not sure if we should explicitly reference `?storage` since ILM will be the norm and there is no `?storage` option there.
```suggestion
are two options for mounting a snapshot, each with different performance
characteristics and local storage footprints:
```

Reviewer 2: Yeah it's tricky, there are many options for mounting a snapshot, but we are only talking about this one parameter. That said, re-reading this after a break I prefer your wording.

[[full-copy]]
Full copy::
Loads a full copy of the snapshotted index's shards onto node-local storage

> within the cluster. This is the default mount option.

Reviewer: I'd point out the default per ILM phase.
```suggestion
within the cluster. This is the default mount option. {ilm-init} uses this
option by default in the `hot` and `cold` phases.
```

+
Search performance for a full-copy searchable snapshot index is normally
comparable to a regular index, since there is minimal need to access the
snapshot repository. While recovery is ongoing, search performance may be
slower than with a regular index because a search may need some data that has
not yet been retrieved into the local copy. If that happens, {es} will eagerly
retrieve the data needed to complete the search in parallel with the ongoing
recovery.

[[shared-cache]]
Shared cache::
+
experimental::[]
+
Uses a local cache containing only recently searched parts of the snapshotted

> index's data. {ilm-init}-managed indices use this option by default when they
> move into the frozen tier.

Reviewer: I'd point out the ILM phase.
```suggestion
index's data. {ilm-init} uses this option by default in the `frozen` phase and
corresponding frozen tier.
```

+
If a search requires data that is not in the cache, {es} fetches the missing
data from the snapshot repository. Searches that require these fetches are
slower, but the fetched data is stored in the cache so that similar searches

> can be served more quickly in future. {es} will evict infrequently-used data

Reviewer: Nit
```suggestion
can be served more quickly in future. {es} will evict infrequently used data
```

=== How {search-snaps} work

When an index is mounted from a snapshot, {es} allocates its shards to data
nodes within the cluster. The data nodes then automatically retrieve the
relevant shard data from the repository onto local storage, based on the mount
options specified. Searches use the data held in local storage and, if not
locally available, download the necessary data from the repository. This avoids
incurring the cost or performance penalty associated with reading data from the
repository when the data is locally available.

If a node holding one of these shards fails, {es} automatically allocates it to

> another node, and that node restores the relevant shard data again from the

Reviewer 3: ```suggestion
another node, and that node restores the relevant shard data from the
```

repository. No replicas are needed, and no complicated monitoring or orchestration
is necessary to restore lost shards.

> In the "full copy" mode, {es} restores a full copy of the {search-snap} shards

Reviewer 3: ```suggestion
In the _full copy_ mode, {es} restores a full copy of the {search-snap} shards
```

in the background and you can search them even if they have not been fully
restored just yet. If a search hits such a {search-snap} shard before it has been
fully restored, {es} eagerly retrieves the data needed to complete the search.

> In the "shared cache" mode, {es} only stores small parts of the data locally on

Reviewer 3: ```suggestion
In the _shared cache_ mode, {es} only stores small parts of the data locally on
```

a node in the cluster. This node-local shared cache has a fixed size and evicts

> mapped file parts based on a "least-frequently-used" policy.

Reviewer 3: ```suggestion
mapped file parts based on a least-frequently-used policy.
```

The size of the space reserved for the shared cache. Defaults to `0b`, meaning
that the node has no shared cache.

For example:

[source,yaml]
----
xpack.searchable.snapshot.shared_cache.size: "4TB"
----

IMPORTANT: Currently, you can configure
`xpack.searchable.snapshot.shared_cache.size` on any node. In a future release,
you will only be able to configure this setting on nodes with the
<<data-frozen-node,frozen data>> role.

You can set `xpack.searchable.snapshot.shared_cache.size` to any size between a
couple of gigabytes up to 90% of available disk space. We only recommend higher
sizes you use the node exclusively as a frozen tier or for searchable snapshots.

[discrete]
[[searchable-snapshots-frozen-tier-on-cloud]]
==== Configure a frozen tier on the {ess}

The frozen data tier is not yet available on the {ess-trial}[{ess}]. However,
you can configure another tier to use <<shared-cache,shared snapshot caches>>.
This effectively recreates a frozen tier in your {ess} deployment. Follow these
steps:

. Choose an existing tier to use. Typically, you'll use the cold tier, but the
hot and warm tiers are also supported. You can use this tier as a shared tier, or

> you can dedicate the tier exclusively to the shared snapshot cache.

Reviewer: I felt that we should give some kind of recommendation here. Using the cold tier made sense to me, but I wasn't sure how well this would play with ILM defaults.

experimental::[]

Using {search-snap} in the "shared cache" mode, where only parts of the
data are locally cached on the nodes in the cluster, requires configuring a
shared snapshot cache which is used to hold a copy of just the
frequently-accessed parts of shards of indices which are mounted with
`?storage=shared_cache`. The `shared_cache` storage option is for example used
by the <<ilm-searchable-snapshot,ILM searchable snapshot action>> in the

> <<frozen-tier,frozen tier>>.

Reviewer 3: ```suggestion
Using {search-snap} in the shared cache mode, where only parts of the
data are locally cached on the nodes in the cluster, requires configuring a
shared snapshot cache. The shared snapshot cache is used to hold a copy of just the
frequently-accessed parts of shards of indices, which are mounted with
`?storage=shared_cache`. The `shared_cache` storage option is used
by the <<ilm-searchable-snapshot,ILM searchable snapshot action>> in the
<<frozen-tier,frozen tier>>.
```

If you configure a node to have a shared cache (disabled by default) then

> that node will fully reserve the specified amount of space for the cache at

Reviewer 3: ```suggestion
that node reserves the specified amount of space for the cache at
```

start up. Indices mounted with the `shared_cache` option can only be allocated
to nodes that have the shared cache explicitly configured.

`xpack.searchable.snapshot.shared_cache.size`::
(<<static-cluster-setting,Static>>, <<byte-units,byte value>>)
The size of the space reserved for the shared cache. Defaults to `0b`, meaning
that the node has no shared cache.

A reasonable value for the shared cache is anything between a couple of
gigabytes up to 90% of available disk space, if the node is to be exclusively

> used for indices mounted with the `shared_cache` option.

Reviewer 3: ```suggestion
gigabytes up to 90% of available disk space if the node is exclusively
used for indices mounted with the `shared_cache` option.
```

NOTE: The shared cache can currently be configured on any node. In the future
this will be restricted to nodes having the <<frozen-tier,frozen tier>> data
role.

[discrete]
[[searchable-snapshots-shared-cache-ex]]
==== Example

Configuring a shared cache that can hold up to 4 terabytes of data is done by

> adding the following line to your `elasticsearch.yml` file:

Reviewer 3: ```suggestion
To configure a shared cache that can hold up to 4 terabytes of data,
add the following entry to `elasticsearch.yml`:
```

[source,yaml]
--------------------------------------------------
xpack.searchable.snapshot.shared_cache.size: "4TB"
--------------------------------------------------

[discrete]
[[searchable-snapshots-frozen-tier-on-cloud]]
==== Configuring the shared snapshot cache for the frozen tier on {ess}

On {ess}, the frozen tier is not fully integrated yet and requires a simple
manual configuration step.

Users in {ess} will have to chose one of the existing tiers in Cloud
(hot/warm/cold) to run the frozen tier functionality on. This can be configured
by link:{cloud}/ec-add-user-settings.html[adding the `xpack.searchable.snapshot.shared_cache.size` user setting]
to one of the existing tiers in the Elasticsearch Service Console.

Depending on whether the hot/warm/cold tier is to be exclusively used for the
new frozen functionality or whether it is to be shared with other data
on that tier, the shared_cache.size can be configured from just a few

> gigabytes up to 90% of the available disk space.

Reviewer 3: ```suggestion
To use the frozen functionality in {ess}, you currently need to manually configure the shared cache on one of the existing tiers (hot/warm/cold). 

From the Elasticsearch Service console, link:{cloud}/ec-add-user-settings.html[add the `xpack.searchable.snapshot.shared_cache.size` user setting]
to the selected tier.

If the tier is used exclusively as the frozen tier, the cache can be allotted up to 90% of the available disk space. If the tier is shared--for example, holds both cold and frozen data--the shared cache size might be just a few gigabytes.
```
